plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -143,18 +145,27 @@ instance ExMemoryUsage () where
   memoryUsage () = 1
 
 instance ExMemoryUsage Integer where
-  memoryUsage 0 = ExMemory 1  -- integerLog2# is unspecified for 0, but in practice returns -1
-  memoryUsage i = ExMemory . fromIntegral $ (1 + smallInteger (integerLog2# (abs i) `quotInt#` integerToInt 64)) -- Assume 64bit size.
-
+  memoryUsage 0 = ExMemory 1  -- integerLog2# is unspecified for 0 (but in practice returns -1)
+  memoryUsage i = ExMemory $ fromIntegral $ (I# n) + 1
+                               where n = (integerLog2# (abs i) `quotInt#` integerToInt 64) :: Int#
+                               -- Assume 64-bit size for Integer
+
+{- Bytestrings: we want things of length 0 to have size 0, 1-8 to have size 1,
+   9-16 to have size 2, etc.  Note that (-1) div 8 == -1, so the code below
+   gives the correct answer for the empty bytestring.  Maybe we should just use
+   1 + (toInteger $ BS.length bs) `div` 8, which would count one extra for
+   things whose sizes are multiples of 8. -}
 instance ExMemoryUsage BS.ByteString where
-  memoryUsage bs = ExMemory . fromIntegral $ 1 + ((toInteger $ BS.length bs)-1) `quot` 8
--- We want things of length 0-8 to have size 1, 9-16 to have size 2, etc.
--- We use 'quot' to deal with the empty bytestring because 'div' would give -1.
--- Maybe we should just use 1 + (toInteger $ BS.length bs) `div` 8, which
--- would count one extra for things whose sizes are multiples of 8.
+  memoryUsage bs = ExMemory $ ((n-1) `div` 8) + 1
+      where n = fromIntegral $ BS.length bs :: SatInt
 
+{- Text objects are UTF-16 encoded, which uses two bytes per character (strictly,
+   codepoint) for everything in the Basic Multilingual Plane but four bytes for
+   the other planes.  We use lengthWord16 because (a) it tells us the actual number
+   of 2-byte words used, and (2) it's O(1), but T.length is O(n). -}
 instance ExMemoryUsage T.Text where
-  memoryUsage text = memoryUsage $ T.unpack text -- TODO not accurate, as Text uses UTF-16
+  memoryUsage s = ExMemory $ ((n-1) `div` 4) + 1
```

Comment:
I'm not sure if we can just change this, I don't know what will happen if the ledger starts using this. I think it's unsafe since it would affect checking history...

Reply:
I now think this change is safe.  A `Text` object consists of an array of `Word16` values together with a `start` index and a `length` field, which allows you to take a substring efficiently by adjusting the start and length. The `lengthWord16` function returns the length field, so it may actually lie about the size of the object (if you take a 2-character substring of a 100-chaaracter string then `lengthWord16` will return 2, but the underlying object will occupy 200 bytes.  I don't think that's a problem though, since we'll never process characters outside such a substring, and I don't think we can do anything that could cause the whole 200 bytes to be copied when we're just interested in the small substring.  

The new `ExMemoryUsage` instance for `Text` is based on the number returned by `lengthWords16`, which may not correspond exactly to the number of Unicode characters in the string: each Unicode character occupies either one or two 16-bit words in the array (one word for characters in plane 0, two in the other planes), so at one extreme a `Text` object of `lengthWord16` `n` will represent `n` characters and at the other `n/2` characters.  Now the formula just below here says that the memory usage corresponds to the length as follows:
```
|lengthWord16    |   memoryUsage | min chars | max chars |
-----------------------------------------------
| 0 | 0 | 0 | 0 |
|1 | 1 | 1 | 1 |
|2 | 1 | 1 | 2 |
|3 | 1 | 2 | 3 |
|4 | 1 | 2 | 4 | 
|5 | 2 | 3 | 5 |
|6 | 2 | 3 | 6 |
|7 | 2 | 4 | 7 |
|8 | 2 | 4 | 8 |
|9 | 3 | 5 | 9 | 
| 10 | 3 | 5 | 10 | 
|... | ...| ... | ... |

(Remember that `ExMemoryUsage` measures memory in four-byte words, which I'm starting to think was a bad decision: bytes would have been a lot simpler because we wouldn't have to do this division by 4 stuff).

Anyway, the upshot of this is that a `Text` string with memory usage `k` contains at most  `k` characters (when k = 0 or 1) and usually a lot more. The old `ExMemoryUsage` instance converts the `Text` object to a `String`, and then charges one unit (4 bytes) for each character in the string, so in this case a `Text` string with memory usage `k` always contains exactly `k` characters.  So the memory usage of a `k`-character `Text` string according to the new instance is always less than or equal to the usage according to the old instance, and hence any charge based on the new instance will always be less than or equal to the charge according to the new instance (since our costing functions are monotonic), and a transaction that validates successfully with the old instance will always validate successfully with the new instance, and probably with a bit more of the fees left over at the end.

That's quite a complicated chain of reasoning though, so I'd better convince myself completely before proceeding...

Reply:
Mmmm. I guess that is probably okay, but it relies on a particular bit of reasoning that I hadn't considered before: we can change the budgeting approach unilaterally for old things too so long as it is guaranteed to return less than or equal budgets to before.

We should at least check that e.g. the output for the `plutus-use-cases` scripts is the same as before! Fortunately, Hydra produces those now, so it's easy to look.

Reply:
I looked: there are some changes in the CPU usage but not the memory usage. Is that expected?

Reply:
> I looked: there are some changes in the CPU usage but not the memory usage. Is that expected?

The memory usage costs only depend on the string size for `appendString` and `encodeUtf8`, so there might be changes for those but not for others.  Grepping the PLC source of the `plutus-use-cases` scripts suggests that none of them use any of the string functions though.  

I did change quite a few of the memory models in this PR, so some of the memory costs should change (and increase in some cases).

Reply:
In fact I changed the memory models for `appendString` and `encodeUtf8`, so memory usage costs for calling them should be different because both the models and the size function have changed.

Reply:
Okay, happy to merge if you're confident this doesn't change anything.